tests/test_socket_server.py

```diff
@@ -0,0 +1,180 @@
+
+    async def asyncTearDown(self) -> None:
+        await self.srv.disconnect()
+        if self.writer:
+            self.writer.close()
+            await self.writer.wait_closed()
+        await self.srv.exit()
+
+    async def read(self) -> typing.Dict[str, typing.Any]:
+        """Read a string from the reader and unmarshal it.
+
+        Returns
+        -------
+        data : `dict`
+            A dictionary with objects representing the string read.
+        """
+        read_bytes = await asyncio.wait_for(
+            self.reader.readuntil(tcpip.TERMINATOR), timeout=TIMEOUT
+        )
+        data = json.loads(read_bytes.decode())
+        return data
+
+    async def write(self, **data: typing.Dict[str, typing.Any]) -> None:
+        """Write the data appended with a tcpip.TERMINATOR string.
+
+        Parameters
+        ----------
+        data:
+            The data to write.
+        """
+        st = json.dumps({**data})
+        assert self.writer is not None
+        self.writer.write(st.encode() + tcpip.TERMINATOR)
+        await self.writer.drain()
+
+    async def test_disconnect(self) -> None:
+        self.assertTrue(self.srv.connected)
+        await self.write(command=common.Command.DISCONNECT, parameters={})
+        # Give time to the socket server to clean up internal state and exit.
+        await asyncio.sleep(0.5)
+        self.assertFalse(self.srv.connected)
+
+    async def test_exit(self) -> None:
+        self.assertTrue(self.srv.connected)
+        await self.write(command=common.Command.EXIT, parameters={})
+        # Give time to the socket server to clean up internal state and exit.
+        await asyncio.sleep(0.5)
+        self.assertFalse(self.srv.connected)
+
+    def check_temperature_reply(
+        self, reply: typing.List[typing.Union[str, float]]
+    ) -> None:
+        device_name = reply[0]
+        time = reply[1]
+        response_code = reply[2]
+        resp = reply[3:]
+
+        self.assertEqual(self.name, device_name)
+        self.assertGreater(time, 0)
+        if self.in_error_state:
+            self.assertEqual(common.ResponseCode.DEVICE_READ_ERROR, response_code)
+        else:
+            self.assertEqual(common.ResponseCode.OK, response_code)
+        self.assertEqual(len(resp), self.num_channels)
+        for i in range(0, self.num_channels):
+            self.assertLessEqual(common.MockTemperatureConfig.min, resp[i])
+            self.assertLessEqual(resp[i], common.MockTemperatureConfig.max)
+
+    async def check_server_test(self) -> None:
+        """Test a full command sequence of the SocketServer.
+
+        The sequence is
+            - configure
+            - start
+            - read telemetry
+            - stop
+            - disconnect
+            - exit
+        """
+        configuration = {
+            common.Key.DEVICES: [
+                {
+                    common.Key.NAME: self.name,
+                    common.Key.CHANNELS: self.num_channels,
```

Comment:
Please pass num_channels as an argument.

In general it's best to define and document all attributes of a test class in `setUp` and of a normal class in `__init__`. Then other methods can set or modify them as needed. Doing anything else makes the logic harder to follow: where did the variable come from? What does it do? 

There are exceptions. Sometimes it's useful to have a few clearly defined methods whose primary job is to create and add important attributes. salobj BaseCscTestCase.make_csc is an example.

Reply:
I made `num_channels` an argument and improved the documentation.